### DISCUSSION TITLE
Modeling the operation of a cell from real world energy storage system data

### DISCUSSION THREAD

**Discussion by jhthompson12** [pybamm-team/PyBaMM#4147] [Q&A] [answered]
## Background
Hi there, Im a stationary energy storage researcher in the electric utility space and I am very interested in PyBaMM, though much of the electrochemistry and physics in these models goes a bit over my head.

I have 6 years of data from a large storage system (think large shipping containers filled with racks, filled with modules, filled with cells) where we have developed our own characterization of the degradation of this system over the years by analyzing periodic reference performance test data. 

I would like to use some or all of this data (mainly historical current and temperature timeseries) as the input to a PyBaMM model to see how these models show degradation under the same cycling and ambient conditions as the real system. 

So far I have had luck getting everything installed and am working with just a 1 hour snippet of my real system current data. Im trying to input this to a simple NMC model and, if everything worked out, I am hoping that PyBaMM would roughly model the same operation that I see in the real data for this 1-hour time period. 

Unfortunately, the simulation only makes it about 30 minutes into the cycling before stopping, it does not error out, but the simulation data just ends at around 1700 seconds. 

### Some details about the real system I am trying to model
- I have 10 second resolution data at the rack level for this system. Each Rack consists of 17 module. Each Module contains 28 LG Chem JH3 cells in a 14S2P configuration
- To get the cell-level current from the rack-level current data that I have I just divide rack-level current by 2 before handing the current profile off to the PyBaMM model
- I am using the `Mohtat2020` parameters as a starting point because it is listed as NMC, which is the same as my real cells. Then I adjust some of those parameters to match specs of the JH3 cells

### My Script
``` python 
import pybamm
import matplotlib as mpl
import pandas as pd  # needed to read the csv data file
import plotly.graph_objs as go
from plotly.offline import plot
from plotly.subplots import make_subplots

mpl.use('TkAgg')  # or can use 'TkAgg', whatever you have/prefer

# import drive cycle data from file
all_data = pd.read_pickle("../Create Sample Data/Data/2017-2018_Rack1_Data.pickle")

# extract smaller portion of data
start = pd.to_datetime("2017-09-22 00:00:00")
end = pd.to_datetime("2017-09-22 01:00:00")

smaller_data = all_data.loc[start:end][[
    'Container 1, Rack 1_StateOfCharge',
    'Container 1, Rack 1_DC.Voltage_Avg of Cells',
    'Container 1, Rack 1_Temp.Battery.Module_Avg of Modules',
    'Container 1, Rack 1_DC.Current'
]]

# The rack from which this data originates is made of 17 modules (in series), where each module contains 28 cells in a
# 14S2P configuration. So, to get the cell-level current we divide the rack-level current by 2
smaller_data["Cell Current"] = smaller_data['Container 1, Rack 1_DC.Current'] / -2  # the sign convention might be backwards, so flip it

# create a seconds column
smaller_data["Seconds"] = (smaller_data.index - start).total_seconds()

drive_cycle = smaller_data[["Seconds", "Cell Current"]].to_numpy()

# Instantiate a simple model
model = pybamm.lithium_ion.DFN()

# load parameter values
param = model.default_parameter_values

# get predefined parameters for an NMC cell
param_mohtat = pybamm.ParameterValues("Mohtat2020")

# apply the SoC / Voltage relationship data from analyzing the real system data
param_mohtat["Open-circuit voltage at 0% SOC [V]"] = 3.2
param_mohtat["Open-circuit voltage at 100% SOC [V]"] = 4.17
param_mohtat["Upper voltage cut-off [V]"] = 4.2
# nominal capacity for JH3 cells (and a few other parameters) from LG Chem documentation
param_mohtat["Nominal cell capacity [A.h]"] = 63
param_mohtat["Cell volume [m3]"] = 5.65e-04
param_mohtat["Lower voltage cut-off [V]"] = 3

# create interpolant - must be a function of *dimensional* time
current_interpolant = pybamm.Interpolant(drive_cycle[:, 0], drive_cycle[:, 1], pybamm.t)

# set drive cycle
param_mohtat["Current function [A]"] = current_interpolant

output_variables = [
    'Voltage [V]',
    'Current [A]',
]

# set up simulation - for drive cycles we recommend using the CasadiSolver in "fast" mode
solver = pybamm.CasadiSolver(mode="fast")
simulation = pybamm.Simulation(model, parameter_values=param_mohtat)

# simulate US06 drive cycle (duration 600 seconds)
initial_soc = smaller_data['Container 1, Rack 1_StateOfCharge'].iloc[0] / 100
simulation.solve(initial_soc=initial_soc, showprogress=True)

solution = simulation.solution

# combine solution data for later plotting
solution_data = pd.concat([pd.Series(solution[var].entries, index=solution.t, name=var) for var in output_variables], axis=1)

# plot
# combine simulation data back into original EW Brown data
plot_data = pd.concat([smaller_data.reset_index().set_index("Seconds"), solution_data], axis=1)
plot_data = plot_data.loc[~plot_data["index"].isna()]

fig = make_subplots(rows=2, cols=1, shared_xaxes=True)

fig.add_trace(
    go.Scattergl(
        x=plot_data["index"],
        y=plot_data['Current [A]'],
        name='Cell Current (Rack Current / 2)',
        mode="lines",
        line=dict(width=1)
    ), row=1, col=1
)

fig.add_trace(
    go.Scattergl(
        x=plot_data["index"],
        y=plot_data['Container 1, Rack 1_DC.Voltage_Avg of Cells'],
        name='Rack Avg Cell Voltage',
        mode="lines",
        line=dict(width=1)
    ), row=2, col=1
)

fig.add_trace(
    go.Scattergl(
        x=plot_data["index"],
        y=plot_data['Voltage [V]'],
        name="Modeled Cell Voltage",
        mode="lines",
        line=dict(width=1)
    ), row=2, col=1
)

fig.update_layout(
    yaxis=dict(title="Cell Current (A)"),
    yaxis2=dict(title="Cell Voltage (V)"),
    # yaxis3=dict(title="AmpHours")
)

plot(fig, filename="Plots/Real_vs_Modeled_Voltage.html")
```

### Simulation Results
![Screenshot 2024-06-06 at 4 16 35 PM](https://github.com/pybamm-team/PyBaMM/assets/45108842/2619b2a1-4fad-4bf9-821c-60c3d16ca2e2)

### Final Questions
- Does what Im attempting to do even make sense? 
- If so, how can I improve upon what I have tried already? Maybe different models? Maybe customizing more of the parameters? 

Thanks in advance for any feedback!

**Comment by TomTranter** [comment]
Hi, it doesn't seem like you're doing anything wrong with the model but it's odd that if you give it a longer time series the sim stops with no error. I didn't have your file but this looks almost right for the current and scripts runs ok

```
import pybamm
import matplotlib.pyplot as plt
import numpy as np

# Define the minute-by-minute values from the table provided
current_values = [
    -1, -3, -5, -4, -3, -1, 1, 3, 4, 2, -1, -3, -5, -7, -9, -10, -8, -6, -4, -3,
    0, 2, 5, 8, 10, 10, 9, 7, 5, 2, -2, -5, -8, -11, -12, -11, -9, -7, -5, -3,
    -1,
]

# Convert the list to a numpy array
current_array = np.array(current_values)
time_array = np.arange(0, len(current_array))*60

plt.figure()
plt.plot(current_array)

# Instantiate a simple model
model = pybamm.lithium_ion.DFN()

# load parameter values
param = model.default_parameter_values

# get predefined parameters for an NMC cell
param_mohtat = pybamm.ParameterValues("Mohtat2020")

# apply the SoC / Voltage relationship data from analyzing the real system data
param_mohtat["Open-circuit voltage at 0% SOC [V]"] = 3.2
param_mohtat["Open-circuit voltage at 100% SOC [V]"] = 4.17
param_mohtat["Upper voltage cut-off [V]"] = 4.2
# nominal capacity for JH3 cells (and a few other parameters) from LG Chem documentation
param_mohtat["Nominal cell capacity [A.h]"] = 63
param_mohtat["Cell volume [m3]"] = 5.65e-04
param_mohtat["Lower voltage cut-off [V]"] = 3

# create interpolant - must be a function of *dimensional* time
current_interpolant = pybamm.Interpolant(time_array, current_array, pybamm.t)

# set drive cycle
param_mohtat["Current function [A]"] = current_interpolant

output_variables = [
    'Voltage [V]',
    'Current [A]',
]


solver = pybamm.CasadiSolver(mode="fast")
simulation = pybamm.Simulation(model, parameter_values=param_mohtat)

simulation.solve(initial_soc=0.5, showprogress=True)
solution = simulation.solution

solution.plot()
```
![image](https://github.com/pybamm-team/PyBaMM/assets/7068741/59579e27-81bd-4307-b005-2ca4525b3454)

**Reply by TomTranter** [reply, marked as answer]
Although changing the nominal cell capacity will have little effect. You need to adjust electrode dimensions to change actual capacity. That's probably why your voltages are changing more than you expect

**Reply by jhthompson12** [reply]
Thank you for the reply! 

I think you're right about the electrode dimensions specfication. I only changed the `Nominal cell capacity [A.h]` to 63 and
`Cell volume [m3]` to 5.65e-04, so i guess the model still effectively has the 5Ah capacity from the `Mohtat2020` parameters, which would definitely explain the larger changes in voltage for a given amount of charge compared to my 63Ah cell. 

Im trying to think through a way to proceed. PyBaMM does not already have a predefined set of parameters for the specific LG Chem JH3 cells for which I have real data. So, I can think of 2 ways forward:
- Option 1: Try to create my own set of custom parameters that closely describe the JH3 cells that I have. 
    - The problem with this is that I have very limited information about the cells internal design, I basically just have the high-level information from public documentation:
        ![image](https://github.com/pybamm-team/PyBaMM/assets/45108842/f347f5c6-baa0-478d-a24a-a2b0f6123e05)

- Option 2: Try just using the most similar predefined parameters from PyBaMM (basically any NMC parameters like `Mohtat2020`) and scale the current such that the modeled cell is experiencing a similar amount of stress from a C-rate perspective.  
    - To create the current that the modeled 5Ah cell should experience I would divide the JH3 current by 63Ah to get the C-rate and then multiply by 5Ah. This would become the current input for the simulation. 

@TomTranter, do you have any thoughts about these approaches? Im going to try option 2 here soon. 

To reiterate, my true end goal is run a few NMC PyBaMM models through simulations using the 6 years of cell current and ambient temperature data that I have from this real system. I would then extract the discharge energy capacity / state of health data from the simulations and compare them to the known degradation that the real system experienced. 

### Bonus Questions! 
Lmk if i should move these to separate Q&A threads.
- The simulation was stopping at around 1700 seconds but printed nothing to the console. I suspect it was stopping because the modeled 5Ah cell reached it's top of charge and was being asked to continue charging. Is there some kind of `verbose` flag for the simulation / solver to be able to better know the status of the simulation? 
- Any thoughts on actually running this model with the full 6 years of 10-second current and temperature data that I have? I could see this being problematic..

**Reply by jhthompson12** [reply]
Option 2 seems to work decently enough, at least at small timescales, like one hour. 

![image](https://github.com/pybamm-team/PyBaMM/assets/45108842/aa4caa68-8db5-43c7-b593-a7c771d64ef5)

Im now running into trouble with longer term simulations (multiple days) where the error between the modeled cell and real cell voltages increases until the modeled cell reaches top of charge stopping the simulation. I think I'll ask a more pointed question in a new thread.

**Reply by TomTranter** [reply]
Are you in the pybamm slack channel? Prof David Howey would be a great person to ask about best approaches for analysing field data with limited info. My instinct is that physics based models might not be the best starting point. Maybe also @valentinsulzer has an opinion

**Reply by TomTranter** [reply]
I just reread that you do have rpt cycles. Have you tried or had any luck fitting the model to any of these?

**Reply by jhthompson12** [reply]
> Are you in the pybamm slack channel?

No, but would love to be! my email is `jhthompson12@gmail.com`

> Have you tried or had any luck fitting the model to any of these?

Im not sure I understand the question. I do have an already made degradation curve from all the RPTs that this system has undergone. My hope for PyBaMM is to find a way to 
- run an NMC simulation over the same 6 years of cell current and ambient temperature data
- extract the model's "remaining discharge capacity" data 
- compare that modeled degradation trend to what the real system shows to see how close the modeled vs reality is 

The larger research question is essentially, "can a physics based modeling tool like PyBaMM or [NREL's BLAST](https://www.nrel.gov/transportation/blast.html) be used to somewhat accurately predict the long-term degradation of a battery system for some arbitrary temperature and duty cycle?"

Right now im stuck at getting the model to work through more than just a few days of my real world current data, but i suspect there is a way to move forward, just need to figure it out!

Thanks!

**Reply by agriyakhetarpal** [reply]
@jhthompson12, an invitation for the PyBaMM Slack channel is available at https://pybamm.org/slack/.

**Reply by davidhowey** [reply]
> Are you in the pybamm slack channel? Prof David Howey would be a great person to ask about best approaches for analysing field data with limited info. My instinct is that physics based models might not be the best starting point. Maybe also @valentinsulzer has an opinion

Thanks for the mention, @TomTranter - I think I agree that a P2D type simulation may not be the best approach here. @jhthompson12 can you elaborate a bit on what your aims are? if it is just to estimate the cell capacity from the real data, then a physical model is overkill and you can use a much simpler model, like an ECM (even just a voltage source and resistor). We have spent some years working on this kind of issue and it's way more challenging than we expected, for many reasons: (1) Parameter estimation is just hard, generally - there is no guarantee that estimates will converge or behave stably. (2) Unmodelled effects mean that there is usually a gap between the model and real world data, especially on a large system like this, so you will need to also worry about tracking the measured data properly (I think you are already facing some of these issues. (3) Issues like OCV hysteresis can really play havoc with things. Happy to chat further.

**Reply by jhthompson12** [reply]
Thanks for the reply, @davidhowey! 

> can you elaborate a bit on what your aims are?

The larger research question is essentially, "can contemporary, readily available battery modeling tools like PyBaMM or [NREL's BLAST](https://www.nrel.gov/transportation/blast.html) be used to somewhat accurately predict the long-term degradation of a battery system for some arbitrary temperature and duty cycle? If not, why?"

So it's really just an exploratory exercise at the moment and I'm hoping to see what happens if I run this large batch of existing data through PyBaMM. Im not too worried about model accuracy at the moment. The deliverable from this effort will probably be a short paper that shows what happens when we run the same cell cycling dataset, with a know degradation trend, through some of these models to see how they model the same degradation. If / when the models do not work very well we will talk about some of the reasons for this, which I think you have a lot of great insight into. 

So my main interests right now are:
1. Finding a way to get a PyBaMM simulation to work through all of this real cycling data. Given [my positive interactions with this community so far](https://github.com/pybamm-team/PyBaMM/discussions/4152#discussioncomment-9707145) Im optimistic that there's nothing really stopping me from doing this, I just need to exorcise my ignorance of the tool to make it happen.  
2. Once I do have a viable PyBaMM workflow set up, i would like to make sure I am making the best decisions possible for things like model selection, parameters, configuration, etc to ensure I am doing PyBaMM justice and am not just getting poor model results because I do not understand what Im doing (which is very likely)

I look forward to hearing more from you all and may take you up on your offer to chat more in the future!

**Reply by davidhowey** [reply]
Thanks for the response! We've done some modelling (i.e., long-term degradation prediction against arbitrary usage) along the lines of what you have discussed, using [this code](https://github.com/Battery-Intelligence-Lab/SLIDE) rather than PyBaMM. We published some of results, which were in the context of a [large grid storage project](https://energysuperhuboxford.org), in [this report](https://energysuperhuboxford.org/wp-content/uploads/2023/06/ESO-Final-Report-V7.pdf) from about page 106 onwards. I'd be happy to email you more details on what we did, or have a quick chat sometime - drop me an email (contact details at https://howey.eng.ox.ac.uk/contact/).
